packages/firestore/src/local/shared_client_state.ts

```diff
@@ -0,0 +1,422 @@
+  activeTargetIds = new SortedSet<TargetId>(primitiveComparator);
+  lastUpdateTime: Date;
+
+  private pendingBatchIds = new SortedSet<BatchId>(primitiveComparator);
+
+  constructor() {
+    this.lastUpdateTime = new Date();
+  }
+
+  get minMutationBatchId(): BatchId | null {
+    return this.pendingBatchIds.first();
+  }
+
+  get maxMutationBatchId(): BatchId | null {
+    return this.pendingBatchIds.last();
+  }
+
+  addPendingMutation(batchId: BatchId): void {
+    assert(
+      !this.pendingBatchIds.has(batchId),
+      `Batch with ID '${batchId}' already pending.`
+    );
+    this.pendingBatchIds = this.pendingBatchIds.add(batchId);
+  }
+
+  removePendingMutation(batchId: BatchId): void {
+    assert(
+      this.pendingBatchIds.has(batchId),
+      `Pending Batch ID '${batchId}' not found.`
+    );
+    this.pendingBatchIds = this.pendingBatchIds.delete(batchId);
+  }
+
+  addQueryTarget(targetId: TargetId): void {
+    assert(
+      !this.activeTargetIds.has(targetId),
+      `Target with ID '${targetId}' already active.`
+    );
+    this.activeTargetIds = this.activeTargetIds.add(targetId);
+  }
+
+  removeQueryTarget(targetId: TargetId): void {
+    assert(
+      this.activeTargetIds.has(targetId),
+      `Active Target ID '${targetId}' not found.`
+    );
+    this.activeTargetIds = this.activeTargetIds.delete(targetId);
+  }
+
+  /** Sets the update time to the current time. */
+  refreshLastUpdateTime(): void {
+    this.lastUpdateTime = new Date();
+  }
+
+  /**
+   * Converts this entry into a JSON-encoded format we can use for LocalStorage.
+   * Does not encode `clientKey` as it is part of the key in LocalStorage.
+   */
+  toLocalStorageJSON(): string {
+    const data: ClientStateSchema = {
+      lastUpdateTime: this.lastUpdateTime.getTime(),
+      activeTargetIds: this.activeTargetIds.toArray(),
+      minMutationBatchId: this.minMutationBatchId,
+      maxMutationBatchId: this.maxMutationBatchId
+    };
+    return JSON.stringify(data);
+  }
+}
+
+/**
+ * `WebStorageSharedClientState` uses WebStorage (window.localStorage) as the
+ *  backing store for the SharedClientState. It keeps track of all active
+ * clients and supports modifications of the current client's data.
+ */
+export class WebStorageSharedClientState implements SharedClientState {
+  private readonly storage: Storage;
+  private readonly storageKey: string;
+  private readonly activeClients: { [key: string]: ClientState } = {};
+  private readonly storageListener = this.handleStorageEvent.bind(this);
+  private readonly clientStateKeyRe: RegExp;
+  private started = false;
+
+  constructor(private persistenceKey: string, private clientKey: string) {
+    if (!WebStorageSharedClientState.isAvailable()) {
+      throw new FirestoreError(
+        Code.UNIMPLEMENTED,
+        'LocalStorage is not available on this platform.'
+      );
+    }
+    this.storage = window.localStorage;
+    this.storageKey = this.toLocalStorageClientKey(this.clientKey);
```

Comment:
storageKey => localClientStorageKey ?  (or remove this and just generate it on-demand from clientKey).

Reply:
Gil suggested this variable in the first PR, which I think makes sense. Renamed.